### PR TITLE
[FIX] fix #1285 #1297 Explicitly model enzymes in OpenMS

### LINKED ISSUE #1285: Enzyme information is missing in XTandem output
If using default setting [RK]|{P} as cleavage site in XTandemAdapter, the output shows
```
	<SearchParameters id="SP_0" db="" charges="+1-+4" enzyme="unknown_enzyme" missed_cleavages="1" precursor_peak_tolerance="1.5" peak_mass_tolerance="0.3" >
```

### PR DESCRIPTION
**Pull request by liangoaix**
fix #1285  and #1297 Explicitly model enzymes in OpenMS

### CODE REVIEW

File: src/topp/XTandemAdapter.cpp [hunk 1]
```diff
@@ -427,6 +427,14 @@ class TOPPXTandemAdapter :
...
     search_parameters.peak_mass_tolerance = getDoubleOption_("fragment_mass_tolerance");
     search_parameters.precursor_tolerance = getDoubleOption_("precursor_mass_tolerance");
+    if (getStringOption_("cleavage_site") == "[RK]|{P}")
+    {
+      search_parameters.enzyme = ProteinIdentification::TRYPSIN;
+    }
+    else if (getStringOption_("cleavage_site") == "[X]|[X]")
+    {
+      search_parameters.enzyme = ProteinIdentification::NO_ENZYME;
+    }
```
Comment (cbielow) [inline src/topp/XTandemAdapter.cpp]: Can we find a way to catch other cases, i.e.
what happens if there is a third regex?
What happens if we have new enzyme supported by OpenMS?

Comment (liangoaix) [inline src/topp/XTandemAdapter.cpp]: I don't know if it's better to use the enzyme names or the specific cleavage sites, not just in XTandem but other cases. In this case, how about adding a cleavage_sites attribute into the SearchParameters, so that a new enzyme called "UNKNOWN_ENZYME" has the cleavage site information.

Comment (timosachsenberg) [inline src/topp/XTandemAdapter.cpp]: from the xtandem documentation "[X]|[X] means cleavage at every peptide bond."
NO_ENZYME would imho suggest there was no cleavage

Comment (liangoaix) [inline src/topp/XTandemAdapter.cpp]: This makes me confused now, I thought it's "no enzyme specified". If it's no cleavage, then should I replace [X]|[X] by unknown_enzyme?

Comment (timosachsenberg) [inline src/topp/XTandemAdapter.cpp]: Good question. I think we don't have something like NON_SPECIFIC_CLEAVAGE.
I think unknown_enzyme would be ok for now but I agree that the enzyme parts in OpenMS should be extended to be more flexible.

**Comment by timosachsenberg** [comment]
@mwalzer do you know of a standardized way to report enzymes in e.g. mzIdentML or related formats that we could use instead of just providing a name?

**Comment by mwalzer** [comment]
Yeah, [X]|[X] is unspecific cleavage (MS:1001956),  ProteinIdentification::NO_ENZYME would be more like NoCleavage (MS:1001955) -  ProteinIdentification::UNKNOWN_ENZYME would be more appropriate.

**Comment by timosachsenberg** [comment]
Do you know how enzymes are specified in mzIdentML? Just putting "trypsin" there doesn't tell us if e.g. proline rule was used.

**Comment by mwalzer** [comment]
```xml
<Enzyme id="ENZ_0" cTermGain="OH" nTermGain="H" missedCleavages="1" semiSpecific="0">
    <SiteRegexp><![CDATA[(?<=[KR])(?!P)]]></SiteRegexp>
    <EnzymeName>
        <userParam name="Enzyme" value="CNBr+Trypsin"/>
    </EnzymeName>
</Enzyme>
```

**Comment by timosachsenberg** [comment]
cool thanks!

**Comment by liangoaix** [comment]
Should we add the same cleavage information here in XTandemAdapter or change the DigestEnzyme Enumerator? @timosachsenberg

**Comment by timosachsenberg** [comment]
I think the best solution is to explicitly model enzymes in OpenMS e.g. see #1297 as a starting point.

**Comment by liangoaix** [comment]
If it's fine so far, I can check the enzyme setting in other search engine adapters.

File: src/openms/source/CHEMISTRY/Enzyme.cpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    regex_description_(enzyme.regex_description_),
+    n_term_gain_(enzyme.n_term_gain_),
+    c_term_gain_(enzyme.c_term_gain_),
+    psi_id_(enzyme.psi_id_),
+    xtandem_id_(enzyme.xtandem_id_),
+    omssa_id_(enzyme.omssa_id_)
+  {
+  }
+
+  Enzyme::Enzyme(const String & name,
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: just a minor thing:
The constructor has the disadvantage that one can create incomplete enzymes. 
We have a similar case with the default constructor, which allows to create an empty enzyme and let the user fill in the details.
Maybe just delete this constructor?

(synonyms_ is also missing here)

File: src/openms/source/CHEMISTRY/Enzyme.cpp [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+#include <OpenMS/CHEMISTRY/Enzyme.h>
+#include <cstdlib>
+#include <iostream>
+
+using namespace std;
+
+namespace OpenMS
+{
+  // enzyme
+  Enzyme::Enzyme() :
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: synonyms_ is missing here

File: src/openms/source/CHEMISTRY/Enzyme.cpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+  ostream & operator<<(ostream & os, const Enzyme & enzyme)
+  {
+    os << enzyme.name_ << " "
+    << enzyme.cleavage_regex_ << " "
+    << enzyme.regex_description_ << " "
+    << enzyme.psi_id_;
+    return os;
+  }
+
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: maybe add operator< so we can store them in odered sets

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+  const Enzyme* EnzymesDB::getEnzyme(const String& name) const
+  {
+    if (enzyme_names_.find(name) == enzyme_names_.end())
+    {
+      throw Exception::ElementNotFound(__FILE__, __LINE__, __PRETTY_FUNCTION__, "Enzyme name cannot be found. '");
+    }
+    return enzyme_names_.at(name);
+  }
+
+  const Enzyme* EnzymesDB::getEnzymebyRegEx(const String & cleavage_regex) const
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: by->By

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+    set<Enzyme*>::iterator it;
+    for (it = enzymes_.begin(); it != enzymes_.end(); ++it)
+    {
+      enzyme_names_[(*it)->getName()] = *it;
+      if ((*it)->getRegEx() != "")
+      {
+        enzyme_regex_[(*it)->getRegEx()] = *it;
+      }
+      set<String>::iterator sit;
+      set<String> syn = (*it)->getSynonyms();
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: I know it is done in some other parts of OpenMS but I would not add the synonyms. Better to have one name as global identifier in OpenMS than several that link to the same enzyme.

**Comment by timosachsenberg** [comment]
Nice work. I have some minor comments but in general it looks good.

File: src/openms/source/CHEMISTRY/Enzyme.cpp (hunk 2 above)
Comment (liangoaix) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: You are right. I will delete this.

File: src/openms/source/CHEMISTRY/Enzyme.cpp (hunk 4 above)
Comment (liangoaix) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: good point. I will do it.

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp (hunk 6 above)
Comment (liangoaix) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: You are right. I was thinking about removing these synonyms, because it's enough to assign one name to each RegEx and each enzyme in OpenMS, there's no need using those synonyms for the same enzyme. The users will only be given the certain options of enzymes from the database.

File: src/openms/source/CHEMISTRY/Enzyme.cpp (hunk 2 above)
Comment (hroest) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: in this case, you should not delete it but make it private

Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: Which one do you mean?

Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: OK. I agree with Hannes that the default constructor should be private and this constr (Name + regex) should be extended to contain parameters for every member. That way we can control that no incomplete enzyme (e.g. no name , etc.) is created.

Comment (liangoaix) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: @timosachsenberg  Do you mean by make Enzyme() private?

Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: Yes exactly.

File: src/openms/source/METADATA/ProteinIdentification.cpp [hunk 7]
```diff
@@ -81,7 +81,8 @@ namespace OpenMS
     enzyme(UNKNOWN_ENZYME),
     missed_cleavages(0),
     peak_mass_tolerance(0.0),
-    precursor_tolerance(0.0)
+    precursor_tolerance(0.0),
+    digestion_enzyme()
```
Comment (liangoaix) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: @timosachsenberg Here it's initialised by using Enzyme::Enzyme(), I don't know if it's still able to make Enzyme::Enzyme() private.

Comment (timosachsenberg) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: You would here use the constructor that contains one parameter for every member and not the default constructor.
e.g. digestion_enzyme("UKNOWN_ENZYME", ...

**Comment by liangoaix** [comment]
@timosachsenberg I updated the constructor settings. Could you check if it's fine now?

**Comment by timosachsenberg** [comment]
thanks alot. looks good from my side. let's wait till after the release with the merge

File: src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    */
+    //@{
+    /// destructor
+    virtual ~EnzymesDB();
+    //@}
+
+    /** @name Accessors
+    */
+    //@{
+    /// returns a pointer to the enzyme with name
```
Comment (cbielow) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h]: can you add
```
@throw Exception::ElementNotFound if enzyme is unkown
```

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+  {
+    set<Enzyme*>::iterator it;
+    for (it = enzymes_.begin(); it != enzymes_.end(); ++it)
+    {
+      delete *it;
+    }
+    enzymes_.clear();
+    enzyme_names_.clear();
+    enzyme_regex_.clear();
+    const_enzymes_.clear();
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: hmm taking a closer look at this we have this pattern in other classes as well:
https://github.com/OpenMS/OpenMS/blob/e8d0ea305916c37b624654c472afab0e47e5c4bb/src/openms/include/OpenMS/CHEMISTRY/ResidueDB.h
I am a bit confused if this is really required to get const access to the elements without compile errors (at least that is what I guess was the motivitation in the original place).

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+  }
+
+  void EnzymesDB::getAllNames(vector<String> & all_names)
+  {
+    all_names.clear();
+    for (set<Enzyme *>::const_iterator it = enzymes_.begin(); it != enzymes_.end(); ++it)
+    {
+      all_names.push_back((*it)->getName());
+    }
+    sort(all_names.begin(),all_names.end());
```
Comment (liangoaix) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: you are right, after adding Enzymes < operator, I forgot this.

File: src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    */
+    //@{
+    typedef std::set<Enzyme *>::iterator EnzymeIterator;
+    typedef std::set<const Enzyme *>::const_iterator EnzymeConstIterator;
+    //@}
+
+    /// this member function serves as a replacement of the constructor
+    inline static EnzymesDB * getInstance()
+    {
+      static EnzymesDB * db_ = 0;
```
Comment (hroest) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h]: I know we have this pattern elsewhere as well but personally, I am in favour of reducing these static objects. I think they are a mess for parallel processing and do a lot of things implicitly we should be doing explicitly. Also, we should not be dealing with raw pointers anymore in OpenMS. 

Especially as the enzyme data structure is so small, why not just make a regular constructor that reads a file and returns an abstraction of that file? If somebody wants to save loading time, he/she can load it once and then pass it into a processing function

this has an additional advantage that users could submit their personalized enzymes.xml on the command line if they wish to do so. if a tool fells this is necessary, it gives the programmer more control and I think its a better pattern

@liangoaix sorry to bring this up here but as I see this as a problem in OpenMS in general, I would like to not introduce new code that uses this static object pattern

@ all others: thoughts on this?

File: src/openms/include/OpenMS/CHEMISTRY/Enzyme.h [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+public:
+
+    /** @name Constructors
+    */
+    //@{
+    /// copy constructor
+    Enzyme(const Enzyme & enzyme);
+
+    /// detailed constructor
+    explicit Enzyme(const String & name,
```
Comment (hroest) [inline src/openms/include/OpenMS/CHEMISTRY/Enzyme.h]: I dont think `explicit` is needed here, I would remove it  ... I think it is just ignored http://stackoverflow.com/questions/1118680/explicit-keyword-on-multi-arg-constructor

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp (hunk 9 above)
Comment (hroest) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: I am sure there is a clever casting macro that could resolve this :-) 

see for example here https://github.com/hroest/OpenMS/blob/feature/replace_DataProcessing/src/openms/include/OpenMS/CONCEPT/Helpers.h#L50 where we have to cast `std::vector<boost::shared_ptr<T> ` to `std::vector<boost::shared_ptr<const T>` why not do the same thing here (without the boost in the middle)?

File: src/openms/include/OpenMS/CHEMISTRY/Enzyme.h (hunk 12 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/CHEMISTRY/Enzyme.h]: if somebody changes const String & cleavage_regex, to const String & cleavage_regex = "" (or the regex for unspecific cleavage) as default then there is implicit conversion as it behaves as a single parameter constructor. I would prefer to put explicit there but I guess in this case it is a matter of taste.

File: src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h (hunk 11 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h]: I definitly agree that we should tackle this issue in OpenMS. What about we will accept the pull request as is, come up with a (if possible) consistent way for the other static objects and refactor them consistently?

Comment (hroest) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h]: sounds good, no need to put it all on @liangoaix to fix, this should be our issue

Comment (cbielow) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h]: I'm not sure that getting rid of the pattern in general is THE solution.
It depends on what data handled. For enzymes I currently do not see the 
need to enable user control - if that were required, then yes, Hannes' 
solution seems more approriate.

Threading is only a problem if we need write access (which occurred for 
Residues if I remember correctly, so there a new pattern would make 
total sense).

Last point: the current pattern requires a minimal amount of code and we 
do not need to pass around parameters down the function hierarchy 
because it can be accessed from anywhere (read only). I know that some 
people now think global variables are bad, but for some applications 
they are convenient and avoid lots of boilerplate code.
But this double bookkeeping of const/non-const needs to go IMHO.

Am 4/2/2015 um 3:10 PM schrieb Timo Sachsenberg:

File: src/openms/include/OpenMS/CHEMISTRY/Enzyme.h (hunk 12 above)
Comment (hroest) [inline src/openms/include/OpenMS/CHEMISTRY/Enzyme.h]: thats true, so leave as is

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+  {
+    clearEnzymes_();
+    readEnzymesFromFile_(file_name);
+    buildEnzymeNames_();
+  }
+
+  void EnzymesDB::addEnzyme(const Enzyme& enzyme)
+  {
+    Enzyme* r = new Enzyme(enzyme);
+    addEnzyme_(r);
```
Comment (hroest) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: why copy the enzyme here? why not

```addEnzyme_(&enzyme); ```

that should work as well ? are you changing the enzyme? see also Chris' comment below

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+    if (enzyme_regex_.has(cleavage_regex))
+    {
+      return true;
+    }
+    return false;
+  }
+
+  bool EnzymesDB::hasEnzyme(const String& enzy_name) const
+  {
+    if (enzyme_names_.find(enzy_name) != enzyme_names_.end())
```
Comment (hroest) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: ```return (enzyme_names_.find(enzy_name) != enzyme_names_.end()```

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+    {
+      return true;
+    }
+    return false;
+  }
+  
+  bool EnzymesDB::hasEnzyme(const Enzyme* enzyme) const
+  {
+    if (const_enzymes_.find(enzyme) != const_enzymes_.end() )
```
Comment (hroest) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: same

File: src/openms/source/FORMAT/IdXMLFile.cpp [hunk 16]
```diff
@@ -456,6 +456,10 @@ namespace OpenMS
       //enzyme
       String enzyme;
       optionalAttributeAsString_(enzyme, attributes, "enzyme");
+      if (EnzymesDB::getInstance()->hasEnzyme(enzyme))
+      {
+        param_.digestion_enzyme = *EnzymesDB::getInstance()->getEnzyme(enzyme);
```
Comment (hroest) [inline src/openms/source/FORMAT/IdXMLFile.cpp]: maybe this is for fixing later, but I dont see the point of using pointers here

File: src/openms/source/METADATA/ProteinIdentification.cpp [hunk 17]
```diff
@@ -81,7 +81,8 @@ namespace OpenMS
     enzyme(UNKNOWN_ENZYME),
     missed_cleavages(0),
     peak_mass_tolerance(0.0),
-    precursor_tolerance(0.0)
+    precursor_tolerance(0.0),
+    digestion_enzyme("unknown_enzyme","")
```
Comment (hroest) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: why treating the enzyome as a string? wouldnt it make sense to here actually store a pointer or an actual `Enzyme` object? or am I missing something?

File: src/openms/source/CHEMISTRY/Enzyme.cpp (hunk 2 above)
Comment (hroest) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: regarding private constructures, I think this may srew up the Python bindings ...

**Comment by hroest** [comment]
@liangoaix great work, thanks for this - and thanks for thinking about the Python bindings :-) did you get  a chance to compile them and see whether they work?  

also a great set of tests, I like it

File: src/openms/source/CHEMISTRY/Enzyme.cpp (hunk 2 above)
Comment (liangoaix) [inline src/openms/source/CHEMISTRY/Enzyme.cpp]: After changing the constructors, I've not yet tested PyOpenMS. report this later

File: src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h (hunk 8 above)
Comment (liangoaix) [inline src/openms/include/OpenMS/CHEMISTRY/EnzymesDB.h]: no problem

File: src/openms/source/CHEMISTRY/EnzymesDB.cpp (hunk 14 above)
Comment (liangoaix) [inline src/openms/source/CHEMISTRY/EnzymesDB.cpp]: thanks for this.

File: src/openms/source/METADATA/ProteinIdentification.cpp (hunk 17 above)
Comment (liangoaix) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: it's making sense to store it as a pointer. I didn't know which one is better and struggled a little bit. Thanks for bringing it up.

Comment (timosachsenberg) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: I think it is a good idea to store an actual enzyme (no pointer).

Comment (hroest) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: But this will be copied many times in memory, right?

sent from my smartphone -- you may keep any typos you find.

Comment (timosachsenberg) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: Do you think this is an issue? I would guess it is OK and simpler than the pointer.

Comment (cbielow) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: An 'Enzyme' is HUGE! Our ID datastructures are already too memory intensive. Please don't make it worse.
A string is uniquely identifyable right?!
Better would be an enum, but since the DB is dynamic I guess thats hardly possible.
Maybe a UID (8byte instead of 24+X)...

Comment (liangoaix) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: yes, it's unique. I think, instead of storing an enzyme, an enzyme name UID could be the solution..

Comment (hroest) [inline src/openms/source/METADATA/ProteinIdentification.cpp]: That is what I feared that copying all the enzyme structures wouldn't be
very.efficient

But then the uid should also go into the xml for ease of use?

sent from my smartphone -- you may keep any typos you find.

**Comment by timosachsenberg** [comment]
The discussion list gets kind of long so let's quickly check what questions need to be resolved and what needs to be done to merge the pull request.
@liangoaix can you go over the comments again and check if they have been addressed and summarize  those that still need some assistance and further discussions?

**Comment by liangoaix** [comment]
The last thing is to store enzyme in ProteinID in an efficient way. I add uid to each enzyme in the enzyme DB and store the uid, is it a good solution?

**Comment by hroest** [comment]
It seems there are 3 good ways: i) human readable designation which was
your initial proposal, ii) a uuid or iii) a pointer

I am fine with either one. It seems you have implemented i) already so we
might as well go with that. As the id is anyways OpenMS specific we dont
really benefit from a uid but it might make sense to use a nomenclature in
i) that is compatible with other people/software.

In conclusion: either of i ii iii is fine and I am fine with the current
implementation as long as there is a unique way of finding the enzyme

sent from my smartphone -- you may keep any typos you find.

**Comment by liangoaix** [comment]
Then I find an easiest way to finish up.

**Comment by timosachsenberg** [comment]
Hi @liangoaix  has everything been adressed and is it good to merge?

**Comment by liangoaix** [comment]
Yes, please merge this. I will carefully replace the schema of const pointer set and normal pointer set storage also in ResidueDB.
